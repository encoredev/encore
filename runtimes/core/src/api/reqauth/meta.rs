use std::str::FromStr;

#[derive(Debug, Copy, Clone, Hash, Eq, PartialEq)]
pub enum MetaKey {
    TraceParent,
    TraceState,
    XCorrelationId,
    Version,
    UserId,
    UserData,
    Caller,
    Callee,
    SvcAuthMethod,
    SvcAuthEncoreAuthHash,
    SvcAuthEncoreAuthDate,
}

impl MetaKey {
    pub fn header_key(&self) -> &'static str {
        use MetaKey::*;
        match self {
            TraceParent => "traceparent",
            TraceState => "tracestate",
            XCorrelationId => "x-correlation-id",
            Version => "x-encore-meta-version",
            UserId => "x-encore-meta-userid",
            UserData => "x-encore-meta-authdata",
            Caller => "x-encore-meta-caller",
            Callee => "x-encore-meta-callee",
            SvcAuthMethod => "x-encore-meta-svc-auth-method",
            SvcAuthEncoreAuthHash => "x-encore-meta-svc-auth",
            SvcAuthEncoreAuthDate => "x-encore-meta-date",
        }
    }
}

pub struct NotMetaKey;

impl FromStr for MetaKey {
    type Err = NotMetaKey;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        use MetaKey::*;
        Ok(match value {
            "traceparent" => TraceParent,
            "tracestate" => TraceState,
            "x-correlation-id" => XCorrelationId,
            "x-encore-meta-version" => Version,
            "x-encore-meta-userid" => UserId,
            "x-encore-meta-authdata" => UserData,
            "x-encore-meta-caller" => Caller,
            "x-encore-meta-callee" => Callee,
            "x-encore-meta-svc-auth-method" => SvcAuthMethod,
            "x-encore-meta-svc-auth" => SvcAuthEncoreAuthHash,
            "x-encore-meta-date" => SvcAuthEncoreAuthDate,
            _ => return Err(NotMetaKey),
        })
    }
}
pub trait MetaMapMut: MetaMap {
    fn set(&mut self, key: MetaKey, value: String) -> anyhow::Result<()>;
}

pub trait MetaMap {
    fn get_meta(&self, key: MetaKey) -> Option<&str>;
    fn meta_values<'a>(&'a self, key: MetaKey) -> Box<dyn Iterator<Item = &'a str> + 'a>;

    /// Returns all meta keys, sorted alphabetically based on MetaKey::header_key.
    fn sorted_meta_keys(&self) -> Vec<MetaKey>;
}

impl MetaMapMut for reqwest::header::HeaderMap {
    fn set(&mut self, key: MetaKey, value: String) -> anyhow::Result<()> {
        self.insert(key.header_key(), value.parse()?);
        Ok(())
    }
}

impl MetaMap for axum::http::HeaderMap {
    fn get_meta(&self, key: MetaKey) -> Option<&str> {
        self.get(key.header_key()).and_then(|v| v.to_str().ok())
    }

    fn meta_values<'a>(&'a self, key: MetaKey) -> Box<dyn Iterator<Item = &'a str> + 'a> {
        Box::new(
            self.get_all(key.header_key())
                .iter()
                .filter_map(|v| v.to_str().ok()),
        )
    }

    fn sorted_meta_keys(&self) -> Vec<MetaKey> {
        let mut keys: Vec<_> = self
            .keys()
            .filter_map(|k| MetaKey::from_str(k.as_str()).ok())
            .collect();
        keys.sort_by_key(|k| k.header_key());
        keys
    }
}

impl MetaMapMut for pingora::http::RequestHeader {
    fn set(&mut self, key: MetaKey, value: String) -> anyhow::Result<()> {
        self.insert_header(key.header_key(), value)?;
        Ok(())
    }
}

impl MetaMap for pingora::http::RequestHeader {
    fn get_meta(&self, key: MetaKey) -> Option<&str> {
        self.headers
            .get(key.header_key())
            .and_then(|v| v.to_str().ok())
    }

    fn meta_values<'a>(&'a self, key: MetaKey) -> Box<dyn Iterator<Item = &'a str> + 'a> {
        Box::new(
            self.headers
                .get_all(key.header_key())
                .iter()
                .filter_map(|v| v.to_str().ok()),
        )
    }

    fn sorted_meta_keys(&self) -> Vec<MetaKey> {
        let mut keys: Vec<_> = self
            .headers
            .keys()
            .filter_map(|k| MetaKey::from_str(k.as_str()).ok())
            .collect();
        keys.sort_by_key(|k| k.header_key());
        keys
    }
}
