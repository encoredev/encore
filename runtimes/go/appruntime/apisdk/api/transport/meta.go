package transport

// These are predefined keys for metadata that we use in Encore.
//
// which allow each transport method know about them and handle them
// in a transport specific way if needed (including renaming them)
const (
	TraceParentKey   = "Traceparent"
	TraceStateKey    = "Tracestate"
	CorrelationIDKey = "Correlation-ID"
)
