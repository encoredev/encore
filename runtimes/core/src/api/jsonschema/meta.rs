use std::collections::HashMap;
use std::ops::Deref;
use std::sync::Arc;

use anyhow::{Context, Result};

use crate::api::jsonschema::de::{Basic, BasicOrValue, Field, Literal, Struct};
use crate::api::jsonschema::{JSONSchema, Registry, Value};
use crate::encore::parser::meta::v1 as meta;
use crate::encore::parser::schema::v1 as schema;
use crate::encore::parser::schema::v1::r#type::Typ;

impl Registry {
    pub fn schema(self: &Arc<Self>, id: usize) -> JSONSchema {
        JSONSchema {
            registry: self.clone(),
            root: id,
        }
    }
}

/// Builder builds a JSONSchema registry.
pub struct Builder<'a> {
    md: &'a meta::Data,

    /// Values that have been computed so far.
    /// None indicates a declaration that's being computed;
    /// it's stored as None until it's computed to be able to handle
    /// recursive references.
    values: Vec<Option<Value>>,

    /// Map of declaration ids to value indices.
    decls: HashMap<u32, usize>,
}

struct BuilderCtx<'a, 'b: 'a> {
    builder: &'a mut Builder<'b>,

    /// The ids of computed type arguments, for the current declaration being processed.
    type_args: &'a [usize],
}

impl<'a> Builder<'a> {
    pub fn new(md: &'a meta::Data) -> Self {
        Self {
            md,
            values: Vec::new(),
            decls: HashMap::new(),
        }
    }

    pub fn build(self) -> Arc<Registry> {
        // Ensure all values have been computed.
        let mut values = Vec::with_capacity(self.values.len());
        for v in self.values {
            values.push(v.expect("missing value"));
        }

        Arc::new(Registry { values })
    }

    #[inline]
    pub fn get(&self, idx: usize) -> Option<&Value> {
        self.values.get(idx).and_then(|v| v.as_ref())
    }

    #[inline]
    pub fn register_value(&mut self, val: Value) -> usize {
        match val {
            // If it's already a ref, return it unmodified.
            Value::Ref(idx) => idx,
            val => {
                let mut ctx = BuilderCtx {
                    builder: self,
                    type_args: &[],
                };
                ctx.reg(val)
            }
        }
    }

    #[inline]
    pub fn register_type(&mut self, typ: &schema::Type) -> Result<usize> {
        let mut ctx = BuilderCtx {
            builder: self,
            type_args: &[],
        };
        let val = ctx.typ(typ)?;

        Ok(match val {
            // If it's a ref, return its index directly.
            Value::Ref(idx) => idx,
            val => ctx.reg(val),
        })
    }

    pub fn struct_field<'b>(&mut self, f: &'b schema::Field) -> Result<(&'b String, Field)> {
        // This should be safe to do because it's only called for schema types,
        // and schema types don't include any type arguments, so we shouldn't need to worry
        // about missing type arguments.
        let ctx = &mut BuilderCtx {
            builder: self,
            type_args: &[],
        };
        ctx.struct_field(f)
    }
}

impl<'a, 'b> BuilderCtx<'a, 'b> {
    /// Computes the JSONSchema value for the given type.
    #[inline]
    fn typ<T: ToType>(&mut self, typ: T) -> Result<Value> {
        let typ = typ.tt()?;

        match typ {
            Typ::Named(named) => self.named(named),
            Typ::Builtin(builtin) => {
                let builtin = schema::Builtin::try_from(*builtin).context("invalid builtin")?;
                Ok(self.builtin(builtin))
            }

            Typ::Pointer(ptr) => self.ptr(ptr),
            Typ::Struct(st) => Ok(Value::Struct(self.struct_val(st)?)),
            Typ::Map(map) => self.map(map),
            Typ::List(list) => self.list(list),
            Typ::Union(union) => self.union(union),
            Typ::Literal(lit) => self.literal(lit),
            Typ::Config(_) => anyhow::bail!("config not yet supported"),
            Typ::TypeParameter(param) => {
                let idx = self
                    .type_args
                    .get(param.param_idx as usize)
                    .ok_or_else(|| anyhow::anyhow!("missing type argument"))?;
                Ok(Value::Ref(*idx))
            }
        }
    }

    #[inline]
    fn named(&mut self, named: &schema::Named) -> Result<Value> {
        let decl = self
            .builder
            .md
            .decls
            .get(named.id as usize)
            .context("missing decl")?;

        // Compute indices for the type arguments.
        let type_args: Result<Vec<usize>> = named
            .type_arguments
            .iter()
            .map(|t| self.typ(t).map(|v| self.reg(v)))
            .collect();
        let type_args = type_args?;

        // Create a nested context that includes the type arguments.
        let mut nested = BuilderCtx {
            builder: self.builder,
            type_args: &type_args,
        };

        let idx = nested.decl(decl)?;
        Ok(Value::Ref(idx))
    }

    #[inline]
    fn decl(&mut self, decl: &schema::Decl) -> Result<usize> {
        // Do we have a value for this decl already?
        if let Some(idx) = self.builder.decls.get(&decl.id) {
            return Ok(*idx);
        }

        // Allocate an index first to handle recursive references.
        let idx = self.builder.values.len();
        self.builder.values.push(None);
        self.builder.decls.insert(decl.id, idx);

        // Then compute the type and update the stored value.
        let typ = self.typ(&decl.r#type)?;
        self.builder.values[idx] = Some(typ);
        Ok(idx)
    }

    #[inline]
    fn ptr(&mut self, ptr: &schema::Pointer) -> Result<Value> {
        self.typ(&ptr.base)
    }

    #[inline]
    fn builtin(&mut self, b: schema::Builtin) -> Value {
        use schema::Builtin;
        Value::Basic(match b {
            Builtin::Any | Builtin::Json => Basic::Any,
            Builtin::Bool => Basic::Bool,
            Builtin::String | Builtin::Bytes | Builtin::Uuid | Builtin::UserId => Basic::String,
            Builtin::Time => Basic::DateTime,

            Builtin::Int
            | Builtin::Uint
            | Builtin::Int8
            | Builtin::Int16
            | Builtin::Int32
            | Builtin::Int64
            | Builtin::Uint8
            | Builtin::Uint16
            | Builtin::Uint32
            | Builtin::Uint64
            | Builtin::Float32
            | Builtin::Float64 => Basic::Number,
        })
    }

    #[inline]
    pub fn struct_val(&mut self, st: &schema::Struct) -> Result<Struct> {
        Ok(Struct {
            fields: {
                let mut map = HashMap::with_capacity(st.fields.len());
                for f in &st.fields {
                    let (k, v) = self.struct_field(f)?;
                    map.insert(k.to_owned(), v);
                }
                map
            },
        })
    }

    #[inline]
    fn struct_field<'c>(&mut self, f: &'c schema::Field) -> Result<(&'c String, Field)> {
        // Note: Our JS/TS support don't include the ability to change
        // the JSON name from the field name, so we use the field name unconditionally.

        let typ = self.typ(&f.typ)?;
        let value = match typ {
            Value::Basic(basic) => BasicOrValue::Basic(basic),
            val => self.bov(val),
        };

        Ok((
            &f.name,
            Field {
                value,
                optional: f.optional,
                name_override: None,
            },
        ))
    }

    #[inline]
    fn map(&mut self, map: &schema::Map) -> Result<Value> {
        // Note: JSON doesn't support anything but string keys,
        // so we don't actually track the key type for the purpose
        // of JSON schemas. Ignore it here.
        let value = self.typ(map.value.tt()?)?;
        Ok(Value::Map(self.bov(value)))
    }

    #[inline]
    fn list(&mut self, list: &schema::List) -> Result<Value> {
        let value = self.typ(list.elem.tt()?)?;
        Ok(Value::Array(self.bov(value)))
    }

    #[inline]
    fn union(&mut self, union: &schema::Union) -> Result<Value> {
        let values: Result<Vec<BasicOrValue>> = union
            .types
            .iter()
            .map(|t| self.typ(t).map(|v| self.bov(v)))
            .collect();
        Ok(Value::Union(values?))
    }

    #[inline]
    fn literal(&mut self, literal: &schema::Literal) -> Result<Value> {
        Ok(match literal.value.clone() {
            Some(schema::literal::Value::Str(val)) => Value::Literal(Literal::Str(val)),
            Some(schema::literal::Value::Boolean(val)) => Value::Literal(Literal::Bool(val)),
            Some(schema::literal::Value::Int(val)) => Value::Literal(Literal::Int(val)),
            Some(schema::literal::Value::Float(val)) => Value::Literal(Literal::Float(val)),
            Some(schema::literal::Value::Null(_)) => Value::Basic(Basic::Null),
            None => anyhow::bail!("missing literal value"),
        })
    }

    #[inline]
    fn bov(&mut self, value: Value) -> BasicOrValue {
        match value {
            Value::Basic(basic) => BasicOrValue::Basic(basic),
            val => BasicOrValue::Value(self.reg(val)),
        }
    }

    #[inline]
    fn reg(&mut self, value: Value) -> usize {
        let idx = self.builder.values.len();
        self.builder.values.push(Some(value));
        idx
    }
}

trait ToType {
    fn tt(&self) -> Result<&Typ>;
}

impl<T> ToType for Option<T>
where
    T: ToType,
{
    fn tt(&self) -> Result<&Typ> {
        self.as_ref().context("missing type")?.tt()
    }
}

impl<T> ToType for Box<T>
where
    T: ToType,
{
    fn tt(&self) -> Result<&Typ> {
        self.deref().tt()
    }
}

impl ToType for schema::Type {
    fn tt(&self) -> Result<&Typ> {
        self.typ.as_ref().context("missing type")
    }
}

impl ToType for Typ {
    fn tt(&self) -> Result<&Typ> {
        Ok(self)
    }
}

impl<T: ToType> ToType for &T {
    fn tt(&self) -> Result<&Typ> {
        (*self).tt()
    }
}
