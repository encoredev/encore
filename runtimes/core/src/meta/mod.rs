use serde::Serialize;

use crate::encore::parser::meta::v1 as meta;
use crate::encore::runtime::v1 as rt;

#[derive(Debug, Clone, Default, Serialize)]
pub struct AppMeta {
    /// The Encore application ID. If the application is not linked to the Encore platform this will be an empty string.
    /// To link to the Encore platform run `encore app link` from your terminal in the root directory of the Encore app.
    pub app_id: String,

    /// The base URL which can be used to call the API of this running application.
    ///
    /// For local development it is "http://localhost:<port>", typically "http://localhost:4000".
    ///
    /// If a custom domain is used for this environment it is returned here, but note that
    /// changes only take effect at the time of deployment while custom domains can be updated at any time.
    pub api_base_url: String,

    /// Information about the environment the app is running in.
    pub environment: EnvironmentMeta,

    /// Information about the build.
    pub build: BuildMeta,

    /// Information about the deployment.
    pub deploy: DeployMeta,
}

impl AppMeta {
    pub fn new(rt: &rt::RuntimeConfig, md: &meta::Data) -> AppMeta {
        let env = rt.environment.as_ref();

        let app_id = env.map(|e| e.app_id.clone()).unwrap_or_default();

        let api_base_url = rt
            .infra
            .as_ref()
            .and_then(|infra| infra.resources.as_ref())
            .and_then(|res| res.gateways.first())
            .map(|gw| gw.base_url.clone())
            .unwrap_or_default();

        AppMeta {
            app_id,
            api_base_url,
            environment: env.map_or_else(EnvironmentMeta::default, EnvironmentMeta::from),
            build: BuildMeta::from(md),
            deploy: rt
                .deployment
                .as_ref()
                .map_or_else(DeployMeta::default, DeployMeta::from),
        }
    }
}

#[derive(Debug, Clone, Default, Serialize)]
pub struct EnvironmentMeta {
    /// The name of environment that this application.
    /// For local development it is "local".
    pub name: String,

    /// The type of environment is this application running in.
    /// For local development it is "development".
    pub r#type: EnvironmentType,

    /// The cloud this is running in.
    /// For local development it is "local".
    pub cloud: CloudProvider,
}

impl From<&rt::Environment> for EnvironmentMeta {
    fn from(env: &rt::Environment) -> Self {
        let env_type = rt::environment::Type::try_from(env.env_type)
            .unwrap_or(rt::environment::Type::Unspecified);
        let cloud = rt::environment::Cloud::try_from(env.cloud)
            .unwrap_or(rt::environment::Cloud::Unspecified);
        EnvironmentMeta {
            name: env.env_name.clone(),
            r#type: EnvironmentType::from(&env_type),
            cloud: CloudProvider::from(&cloud),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
#[derive(Default)]
pub enum EnvironmentType {
    // A production environment.
    Production,
    // A long-lived cloud-hosted, non-production environment, such as test environments, or local development.
    #[default]
    Development,
    // A short-lived cloud-hosted, non-production environments, such as preview environments
    Ephemeral,
    // When running automated tests.
    Test,
}

impl From<&rt::environment::Type> for EnvironmentType {
    fn from(env: &rt::environment::Type) -> Self {
        use rt::environment::Type::*;
        match env {
            Production => Self::Production,
            Ephemeral => Self::Ephemeral,
            Test => Self::Test,
            Development | Unspecified => Self::Development,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
#[derive(Default)]
pub enum CloudProvider {
    AWS,
    GCP,
    Azure,
    Encore,
    #[default]
    Local,
}

impl From<&rt::environment::Cloud> for CloudProvider {
    fn from(cloud: &rt::environment::Cloud) -> Self {
        use rt::environment::Cloud::*;
        match cloud {
            Aws => Self::AWS,
            Gcp => Self::GCP,
            Azure => Self::Azure,
            Encore => Self::Encore,
            Local | Unspecified => Self::Local,
        }
    }
}

#[derive(Debug, Clone, Default, Serialize)]
pub struct BuildMeta {
    // The git commit that formed the base of this build.
    pub revision: String,
    // Whether there were uncommitted changes on top of the commit.
    pub uncommitted_changes: bool,
}

impl From<&meta::Data> for BuildMeta {
    fn from(md: &meta::Data) -> Self {
        BuildMeta {
            revision: md.app_revision.clone(),
            uncommitted_changes: md.uncommitted_changes,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct DeployMeta {
    // The unique id of the deployment. Generated by the Encore Platform.
    pub id: String,
    // The time the deployment was made.
    pub deploy_time: chrono::DateTime<chrono::Utc>,
}

impl From<&rt::Deployment> for DeployMeta {
    fn from(rt: &rt::Deployment) -> Self {
        DeployMeta {
            id: rt.deploy_id.clone(),
            deploy_time: rt
                .deployed_at
                .as_ref()
                .and_then(|d| chrono::DateTime::from_timestamp(d.seconds, d.nanos as u32))
                .unwrap_or_else(chrono::Utc::now),
        }
    }
}

impl Default for DeployMeta {
    fn default() -> Self {
        DeployMeta {
            id: "".into(),
            deploy_time: chrono::Utc::now(),
        }
    }
}
