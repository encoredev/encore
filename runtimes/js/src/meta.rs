use encore_runtime_core::meta;
use napi_derive::napi;

#[napi(object)]
#[derive(Debug, Clone)]
pub struct AppMeta {
    pub app_id: String,
    pub api_base_url: String,
    pub environment: EnvironmentMeta,
    pub build: BuildMeta,
    pub deploy: DeployMeta,
}

impl From<meta::AppMeta> for AppMeta {
    fn from(rt: meta::AppMeta) -> Self {
        AppMeta {
            app_id: rt.app_id,
            api_base_url: rt.api_base_url,
            environment: rt.environment.into(),
            build: rt.build.into(),
            deploy: rt.deploy.into(),
        }
    }
}

#[napi(object)]
#[derive(Debug, Clone)]
pub struct EnvironmentMeta {
    pub name: String,
    pub r#type: EnvironmentType,
    pub cloud: CloudProvider,
}

impl From<meta::EnvironmentMeta> for EnvironmentMeta {
    fn from(rt: meta::EnvironmentMeta) -> Self {
        EnvironmentMeta {
            name: rt.name,
            r#type: rt.r#type.into(),
            cloud: rt.cloud.into(),
        }
    }
}

#[napi]
#[derive(Debug)]
pub enum EnvironmentType {
    Production,
    Development,
    Ephemeral,
    Test,
}

impl From<meta::EnvironmentType> for EnvironmentType {
    fn from(rt: meta::EnvironmentType) -> Self {
        match rt {
            meta::EnvironmentType::Production => Self::Production,
            meta::EnvironmentType::Development => Self::Development,
            meta::EnvironmentType::Ephemeral => Self::Ephemeral,
            meta::EnvironmentType::Test => Self::Test,
        }
    }
}

#[napi]
#[derive(Debug)]
#[allow(clippy::upper_case_acronyms)]
pub enum CloudProvider {
    AWS,
    GCP,
    Azure,
    Encore,
    Local,
}

impl From<meta::CloudProvider> for CloudProvider {
    fn from(rt: meta::CloudProvider) -> Self {
        match rt {
            meta::CloudProvider::AWS => Self::AWS,
            meta::CloudProvider::GCP => Self::GCP,
            meta::CloudProvider::Azure => Self::Azure,
            meta::CloudProvider::Encore => Self::Encore,
            meta::CloudProvider::Local => Self::Local,
        }
    }
}

#[napi(object)]
#[derive(Debug, Clone)]
pub struct BuildMeta {
    pub revision: String,
    pub uncommitted_changes: bool,
}

impl From<meta::BuildMeta> for BuildMeta {
    fn from(rt: meta::BuildMeta) -> Self {
        BuildMeta {
            revision: rt.revision,
            uncommitted_changes: rt.uncommitted_changes,
        }
    }
}

#[napi(object)]
#[derive(Debug, Clone)]
pub struct DeployMeta {
    pub id: String,
    pub deploy_time: String,
}

impl From<meta::DeployMeta> for DeployMeta {
    fn from(rt: meta::DeployMeta) -> Self {
        DeployMeta {
            id: rt.id,
            deploy_time: rt.deploy_time.to_string(),
        }
    }
}
