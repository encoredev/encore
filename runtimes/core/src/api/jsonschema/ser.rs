use serde::{Serializer, Serialize, ser::SerializeMap};
use crate::api::jsonschema::{JSONSchema, Struct};
use crate::api::schema::JSONPayload;

struct SchemaSerializeWrapper<'a>
{
    schema: &'a Struct,
    payload: &'a JSONPayload,
}

impl<'a> Serialize for SchemaSerializeWrapper<'a>
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where S: Serializer
    {
        let mut map = serializer.serialize_map(None)?;
        if let Some(payload) = self.payload {
            for (key, value) in self.schema.fields.iter() {
                if let Some(value) = payload.get(key) {
                    map.serialize_entry(key, value)?;
                }
            }
        }
        map.end()
    }
}

impl JSONSchema {
    pub fn to_json(&self, payload: &JSONPayload) -> serde_json::Result<String> {
        serde_json::to_string(&self.serialize(payload))
    }

    pub fn to_json_pretty(&self, payload: &JSONPayload) -> serde_json::Result<String> {
        serde_json::to_string_pretty(&self.serialize(payload))
    }

    pub fn to_vec(&self, payload: &JSONPayload) -> serde_json::Result<Vec<u8>> {
        serde_json::to_vec(&self.serialize(payload))
    }

    pub fn to_vec_pretty(&self, payload: &JSONPayload) -> serde_json::Result<Vec<u8>> {
        serde_json::to_vec_pretty(&self.serialize(payload))
    }

    pub fn serialize<'a>(&'a self, payload: &'a JSONPayload) -> impl Serialize + 'a {
        SchemaSerializeWrapper {
            schema: self.root(),
            payload,
        }
    }
}