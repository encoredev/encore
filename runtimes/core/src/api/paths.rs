use crate::encore::parser::meta::v1 as meta;
use serde::Serialize;
use std::collections::{HashMap, HashSet};

pub trait Pather {
    type Key;
    type Value;

    fn key(&self) -> Self::Key;
    fn value(&self) -> Self::Value;
    fn path(&self) -> &meta::Path;
}

#[derive(Debug, Serialize, Clone)]
pub struct PathSet<K, V> {
    pub main: HashMap<K, Vec<(V, Vec<String>)>>,
    pub fallback: HashMap<K, Vec<(V, Vec<String>)>>,
}

/// Computes paths to register, grouped by the given key for easier correlation.
pub fn compute<P, K, V>(endpoints: impl Iterator<Item = P>) -> PathSet<K, V>
where
    P: Pather<Key = K, Value = V>,
    K: Eq + Clone + std::hash::Hash,
{
    use crate::encore::parser::meta::v1::path_segment::SegmentType;

    let mut main: HashMap<K, Vec<(V, Vec<String>)>> = HashMap::new();
    let mut fallback: HashMap<K, Vec<(V, Vec<String>)>> = HashMap::new();
    for ep in endpoints {
        let path = ep.path();
        let mut entries = Vec::with_capacity(2);

        // Compute the axum path.
        let mut result = String::new();
        for seg in &path.segments {
            let typ = SegmentType::try_from(seg.r#type).unwrap_or(SegmentType::Literal);
            match typ {
                SegmentType::Literal => {
                    result.push('/');
                    result.push_str(&seg.value)
                }
                SegmentType::Param => {
                    result.push_str("/:");
                    result.push_str(&seg.value)
                }
                SegmentType::Wildcard => {
                    // The wildcard is the last segment.
                    // Axum doesn't match e.g. "/" for "/*wildcard", so we need to register both.
                    result.push_str("/");
                    entries.push(result.clone());

                    result.push('*');
                    result.push_str(&seg.value);
                }
                SegmentType::Fallback => {
                    // Axum doesn't match e.g. "/" for "/*wildcard", so we need to register both.
                    result.push_str("/");
                    entries.push(result.clone());

                    result.push('*');
                    result.push_str(&seg.value);
                }
            }
        }
        entries.push(result);

        let is_fallback = path
            .segments
            .last()
            .map_or(false, |seg| seg.r#type == SegmentType::Fallback as i32);

        let key = ep.key();
        let routes = (ep.value(), entries);
        if is_fallback {
            fallback.entry(key).or_insert_with(Vec::new).push(routes);
        } else {
            main.entry(key).or_insert_with(Vec::new).push(routes);
        }
    }

    // Add paths for TSR redirects.
    {
        for paths in [&mut main, &mut fallback] {
            let path_set: HashSet<String> = HashSet::from_iter(
                paths
                    .values()
                    .flatten()
                    .flat_map(|(_, routes)| routes.iter())
                    .cloned(),
            );
            for entries in paths.values_mut() {
                for (_, endpoint_routes) in entries {
                    let mut tsr_routes = Vec::new();
                    for route in endpoint_routes.iter() {
                        // Is this entry incompatible with TSR?
                        if route == "/" || route.contains("/*") || route.ends_with('/') {
                            continue;
                        }

                        let tsr = format!("{}/", route);
                        if !path_set.contains(&tsr) {
                            tsr_routes.push(tsr);
                        }
                    }
                    endpoint_routes.extend(tsr_routes);
                }
            }
        }
    }

    PathSet { main, fallback }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::encore::parser::meta::v1::path_segment::SegmentType;
    use serde::ser::SerializeStruct;

    #[test]
    fn test_basic() {
        let endpoints = vec![
            ep("one", "a", &[lit("foo")]),
            ep("two", "a", &[lit("foo"), lit("bar")]),
        ];

        let paths = compute(endpoints.into_iter());
        insta::assert_yaml_snapshot!(paths);
    }

    #[test]
    fn test_tsr_conflict() {
        let endpoints = vec![
            ep("one", "a", &[lit("foo"), lit("")]),
            ep("two", "b", &[lit("foo")]),
        ];

        let paths = compute(endpoints.into_iter());
        insta::assert_yaml_snapshot!(paths);
    }

    #[test]
    fn test_wildcard() {
        let endpoints = vec![ep("one", "a", &[wildcard("foo")])];

        let paths = compute(endpoints.into_iter());
        insta::assert_yaml_snapshot!(paths);
    }

    #[test]
    fn test_fallback() {
        let endpoints = vec![ep("one", "a", &[fallback("foo")])];

        let paths = compute(endpoints.into_iter());
        insta::assert_yaml_snapshot!(paths);
    }

    fn path(segs: &[meta::PathSegment]) -> meta::Path {
        meta::Path {
            segments: segs.to_vec(),
            r#type: meta::path::Type::Url as i32,
        }
    }

    fn seg(typ: SegmentType, value: &str) -> meta::PathSegment {
        meta::PathSegment {
            r#type: typ as i32,
            value: value.to_string(),
            value_type: meta::path_segment::ParamType::String as i32,
        }
    }

    fn lit(str: &str) -> meta::PathSegment {
        seg(SegmentType::Literal, str)
    }

    fn param(str: &str) -> meta::PathSegment {
        seg(SegmentType::Param, str)
    }

    fn wildcard(str: &str) -> meta::PathSegment {
        seg(SegmentType::Wildcard, str)
    }

    fn fallback(str: &str) -> meta::PathSegment {
        seg(SegmentType::Fallback, str)
    }

    #[derive(Clone, Debug)]
    struct TestEndpoint {
        name: &'static str,
        key: &'static str,
        path: meta::Path,
    }

    impl Serialize for TestEndpoint {
        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where
            S: serde::Serializer,
        {
            let mut state = serializer.serialize_struct("TestEndpoint", 2)?;
            let path = path_to_str(&self.path);
            state.serialize_field("key", &self.key)?;
            state.serialize_field("path", &path)?;
            state.end()
        }
    }

    fn ep(name: &'static str, key: &'static str, segs: &[meta::PathSegment]) -> TestEndpoint {
        TestEndpoint {
            name,
            key,
            path: path(segs),
        }
    }

    impl Pather for TestEndpoint {
        type Key = String;
        type Value = String;

        fn key(&self) -> Self::Key {
            self.key.to_string()
        }
        fn value(&self) -> Self::Value {
            self.name.to_string()
        }

        fn path(&self) -> &meta::Path {
            &self.path
        }
    }
}

pub fn path_to_str(path: &meta::Path) -> String {
    let mut result = String::new();
    for seg in &path.segments {
        result.push('/');

        use meta::path_segment::SegmentType;
        match SegmentType::try_from(seg.r#type).unwrap() {
            SegmentType::Literal => result.push_str(&seg.value),
            SegmentType::Param => {
                result.push(':');
                result.push_str(&seg.value)
            }
            SegmentType::Wildcard | SegmentType::Fallback => {
                result.push('*');
                result.push_str(&seg.value)
            }
        }
    }

    result
}
