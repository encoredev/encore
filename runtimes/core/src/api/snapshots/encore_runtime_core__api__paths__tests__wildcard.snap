---
source: runtimes/core/src/api/paths.rs
expression: paths
---
main:
  a:
    - - one
      - - /
        - /*foo
fallback: {}
